docs/modules/candid-guide/index.adoc

```diff
@@ -0,0 +1,1098 @@
+
+Furthermore, the set of types is chosen to allow a resonably intuitive mapping to the respective host language (Motoko, Rust, JavaScript, etc.), thus settling on very canonic choices.
+
+[#candid-service-descriptions]
+=== Candid service descriptions
+
+With these candid types, we now can describe a service. Such a service description file can be written by hand, but typically it is generated from the service implementation, as described later.
+
+The simplest service description is
+....
+service : {}
+....
+and describes a service with no public methods. This is not very useful, so lets add a simple method:
+....
+service : {
+  ping : () -> ();
+}
+....
+
+This says that the service supports a public method called `ping`. Method names can be arbitrary strings, and you can quote them (`"method with spaces"`) if they are not plain identifiers.
+
+Methods declare a _sequence_ of arguments and result types. In the case of `ping`, no arguments are passed and none are returned, so the empty sequence `()` is used. We can extend that:
+....
+service : {
+  reverse : (text) -> (text);
+  divMod : (dividend : nat, divisor : nat) -> (div : nat, mod : nat);
+}
+....
+
+The method `reverse` expects a single parameter of type `text` and returns one value of type `text`.
+
+The method `divMod` expects and returns two values, all of type `int`. In its signature, we use the ability to name the parameters and result values.
+
+[CAUTION]
+====
+Naming the parameters or results of a method is purely for documentation purposes, but does not actually change the method’s type, or the values being passed. Parameter and results are identified by their _position_, independent of the name.
+
+In particular, Candid will does not prevent you from changing the type to
+....
+  divMod : (dividend : nat, divisor : nat) -> (mod : nat, div : nat);
+....
+or passing the above `divMod` to a service expeting a method that returns `mod` first.
+
+This is thus very different from named _record_ fields, which are semantically relevant.
+====
+
+Often, multiple methods in a service may refer to the same complex type. In that case, the type can be named and reused multiple times:
+....
+type address = record {
+  street : text;
+  city : text;
+  zip_code : nat;
+  country : text;
+};
+service address_book : {
+  set_address: (name : text, address) -> ();
+  get_address: (name : text) -> (opt address) query;
+}
+....
+
+These type definitions merely name an _existing_ type, they do not define a new type. It does not matter whether you use `address` in the function signature, or write out the records. In orther words, Candid uses _structural_ typing.
```

Comment:
```suggestion
These type definitions merely abbreviate an _existing_ type, they do not define a new type. It does not matter whether you use `address` in the function signature, or write out the records. In other words, Candid uses _structural_ typing.
```
Maybe say more here. Maybe "Two abbreviations with different names but equivalent definitions, describe the same type and are interchangeable."

Reply:
```suggestion
These type definitions merely abbreviate an _existing_ type, they do not define a new type. It does not matter whether you use `address` in the function signature, or write out the records. Also, two abbreviations with different names but equivalent definitions, describe the same type and are interchangeable. In other words, Candid uses _structural_ typing.
```